mesa_geo/geospace.py

```diff
@@ -91,41 +94,43 @@ def agents(self):
         return self._agent_layer.agents
 
     @property
-    def layers(self) -> List[ImageLayer | gpd.GeoDataFrame]:
+    def layers(self) -> List[ImageLayer | RasterLayer | gpd.GeoDataFrame]:
         """
         Return a list of all layers in the Geospace.
         """
         return self._static_layers
 
     @property
-    def bounds(self):
+    def total_bounds(self) -> np.ndarray | None:
         """
         Return the bounds of the GeoSpace in [min_x, min_y, max_x, max_y] format.
         """
-        return self._bounds
+        return self._total_bounds
 
-    def update_bounds(self, new_bounds: List[float]) -> None:
+    def _update_bounds(self, new_bounds: np.ndarray) -> None:
         if new_bounds:
-            if self._bounds:
-                new_min_x = min(self.bounds[0], new_bounds[0])
-                new_min_y = min(self.bounds[1], new_bounds[1])
-                new_max_x = max(self.bounds[2], new_bounds[2])
-                new_max_y = max(self.bounds[3], new_bounds[3])
-                self._bounds = [new_min_x, new_min_y, new_max_x, new_max_y]
+            if self._total_bounds is not None:
+                new_min_x = min(self.total_bounds[0], new_bounds[0])
+                new_min_y = min(self.total_bounds[1], new_bounds[1])
+                new_max_x = max(self.total_bounds[2], new_bounds[2])
+                new_max_y = max(self.total_bounds[3], new_bounds[3])
+                self._total_bounds = np.array(
```

Comment:
Why does this have to be Numpy NDArray? I haven't seen any array operation being used elsewhere for `_total_bounds`, but I imagine it'd be much slower than if it were Python list/tuple. Also I have seen elsewhere that you ended up casting this to a list anyway.

Reply:
The type and format (min_x, min_y, max_x, max_y) are kept consistent with that of GeoDataFrame: [link](https://github.com/geopandas/geopandas/blob/28462f8e7fd893e01be52f444e142a1a00c36f8e/geopandas/array.py#L918), so that `layer.total_bounds` returns the same result regardless it's an ImageLayer, RasterLayer, or GeoDataFrame.

I also made it part of `GeoBase` so the same holds for `GeoAgent` and `GeoSpace`.

Reply:
This seems to be the real reason for it being NDArray: https://github.com/geopandas/geopandas/issues/458.
Though is accessing the total bounds a hot point in the code? Maybe my concern does not even apply at all because the number of the method calls is so rare?

Reply:
The total bounds are mostly computed/accessed when:

1. agents/layers are created and added into GeoSapce, then the total bounds of GeoSpace is updated accordingly
2. at each step, the total bounds are sent to frontend together with other data for visualization

IMHO there shouldn't be any huge performance impact, be it np.ndarray or list/tuple. But of course we would need a profiler to make sure of that.